test/integration/mqtt_paho_binding/mqtt_test.go

```diff
@@ -0,0 +1,140 @@
+package mqtt_paho_binding
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"testing"
+	"time"
+
+	"github.com/eclipse/paho.golang/paho"
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/cloudevents/sdk-go/protocol/mqtt_paho/v2"
+	"github.com/cloudevents/sdk-go/v2/binding"
+	"github.com/cloudevents/sdk-go/v2/event"
+	"github.com/cloudevents/sdk-go/v2/protocol"
+
+	. "github.com/cloudevents/sdk-go/v2/binding/test"
+	. "github.com/cloudevents/sdk-go/v2/test"
+)
+
+const (
+	Broker = "127.0.0.1:1883"
+)
+
+type receiveMsg struct {
+	msg binding.Message
+	err error
+}
+
+func TestSendStructuredMessageToStructured(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+	s, r, c, err := getSenderReceiver(ctx, cancel)
+	require.NoError(t, err)
+	defer c()
+
+	EachEvent(t, Events(), func(t *testing.T, eventIn event.Event) {
+		eventIn = ConvertEventExtensionsToString(t, eventIn)
+		in := MustCreateMockStructuredMessage(t, eventIn)
+
+		sendReceive(t, binding.WithPreferredEventEncoding(ctx, binding.EncodingStructured), s, r, in,
+			func(out binding.Message) {
+				assert.Equal(t, binding.EncodingStructured, out.ReadEncoding())
+				eventOut := MustToEvent(t, ctx, out)
+				AssertEventEquals(t, eventIn, ConvertEventExtensionsToString(t, eventOut))
+			})
+	})
+}
+
+func TestSendBinaryMessageToBinary(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+	s, r, c, err := getSenderReceiver(ctx, cancel)
+	require.NoError(t, err)
+	defer c()
+
+	EachEvent(t, Events(), func(t *testing.T, eventIn event.Event) {
+		eventIn = ConvertEventExtensionsToString(t, eventIn)
+		in := MustCreateMockBinaryMessage(eventIn)
+
+		sendReceive(t, binding.WithPreferredEventEncoding(ctx, binding.EncodingBinary), s, r, in,
+			func(out binding.Message) {
+				assert.Equal(t, binding.EncodingBinary, out.ReadEncoding())
+				eventOut := MustToEvent(t, ctx, out)
+				AssertEventEquals(t, eventIn, ConvertEventExtensionsToString(t, eventOut))
+			})
+	})
+}
+
+func sendReceive(t *testing.T, ctx context.Context, sender protocol.Sender, receiver protocol.Receiver,
+	in binding.Message, outAssert func(binding.Message),
+) {
+	receiveChan := make(chan receiveMsg)
+	defer close(receiveChan)
+	go func() {
+		msg, result := receiver.Receive(ctx)
+		receiveChan <- receiveMsg{msg, result}
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			require.Fail(t, "timeout waiting for event")
+			return
+		case receiveMsg := <-receiveChan:
+			require.NoError(t, receiveMsg.err)
+			outAssert(receiveMsg.msg)
+			return
+		case <-time.After(2 * time.Second):
+			// the receiver mightn't be ready before the sender send the message, so we retry
```

Comment:
nit: this implementation wasn't obvious. consider adding moving the comment below and add a `continue` statement so it's clear you want to retry (loop)

Reply:
also, since you retry anyways, do we need all these sleeps?

Reply:
Yes, I think we need to wait here, otherwise, it will throw the error message like this:
```bash
{"level":"error","ts":1688780919.884921,"logger":"fallback","caller":"client/invoker.go:77","msg":"call to Invoker.Invoke(...) has panicked: send on closed channel",
```